### PR TITLE
Add a colab notebook demonstrating TFLite import process

### PR DESCRIPTION
**Pull request by not-jenni**
Preview: https://colab.research.google.com/github/not-jenni/iree/blob/tflite-colab/colab/tflite_text_classification.ipynb

Demonstrates how to compile and run a TFLite model with IREE, using the [text classification](https://www.tensorflow.org/lite/examples/text_classification/overview) model

Closes https://github.com/google/iree/issues/5589

### CODE REVIEW

**Review by ScottTodd — COMMENTED** [review]
This notebook is 1.24MB, so I can't add inline comments: 
image

that's coming from printing the .mlir which contains large constants:
```mlir
%0 = "tosa.const"() {value = dense<"0x4FA3373DA6259D3D233E10BDFB5A523C2DF4B4B...
```

The largest notebook we have right now is 381KB. This isn't _that_ much larger, but it's still large enough that I'd rather we trim it down (and being able to comment inline during reviews would be nice :D)

**Comment by ScottTodd** [comment]
> This notebook is 1.24MB, so I can't add inline comments:
> image
> 
> that's coming from printing the .mlir which contains large constants:
> 
> ```mlir
> %0 = "tosa.const"() {value = dense<"0x4FA3373DA6259D3D233E10BDFB5A523C2DF4B4B...
> ```
> 
> The largest notebook we have right now is 381KB. This isn't _that_ much larger, but it's still large enough that I'd rather we trim it down (and being able to comment inline during reviews would be nice :D)

So I'd suggest not printing the .mlir. There's also a `-mlir-elide-elementsattrs-if-larger=N` flag that would make the output small enough to print, if you can find a way to weave that into the notebook.

Another general comment (that I could put on a specific line if inline comments were possible): the "Extract model vocab and label metadata" and "Input and output processing" sections are specific to this particular TFLite model, so I'd like to see them deemphasized or streamlined if possible. You could collapse those sections by default as a start (see how https://colab.research.google.com/github/google/iree/blob/main/colab/mnist_training.ipynb collapses certain sections).

**Review by ScottTodd — COMMENTED** [review]
Overall looks great!

A few suggestions that I can't add inline yet :P

* Collapse the license header section
* Add some comments to the "Model prep" section (explaining the intput/output data formats and embedded text files) and consider collapsing it by default. There should be enough context there that people can see if their own different models will need similar treatment, or if they can skip over that section entirely.
* `index = 142` for inserting the export attribute is exceptionally brittle and could use a comment :P. You could also do some regex searching to find it (`func @main` line -> insert before the close brace)
* Consider adding comments about points where the user could exit from Python (saving the .mlir or .vmfb files and using them with the native tools, including in their app, etc.)

**Review by KoolJBlack — CHANGES_REQUESTED** [review]
Comments since I can't inline:

* Under “model prep”. The form for the first box doesn’t show by default. 
* Are you familiar with python Pathlib? https://realpython.com/python-pathlib/ I find its much nicer way of handling files/paths compared to `os.join.etc...` and its completely interoperable with most existing builtins. You handle several paths throughout the colab that could benefit.
* You can add an “overview” section to the TFLIte? 
* For the long string constants, if the compiler elide option doesn’t work, you may find you can also achieve the same effect suing regular expressions and substitution: https://docs.python.org/3/library/re.html#re.sub. You can match on any string between the “dense<...>” for the “tosa.const()” ops and just replace with ellipses. You also won’t have to run the compiler twice (assuming you need the full mlir file in subsequent steps)
* Nit: perhaps rename “ireert” to “iree_rt”?

File: colab/tflite_text_classification.ipynb [hunk 1]
```diff
@@ -0,0 +1,514 @@
+{
+  "nbformat": 4,
+  "nbformat_minor": 0,
+  "metadata": {
+    "colab": {
+      "name": "Copy of tflite_text_classification.ipynb",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:6]: Remove "Copy of" from the document name

```suggestion
      "name": "tflite_text_classification.ipynb",
```

File: colab/tflite_text_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,514 @@
...
+  },
+  "cells": [
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "_VsBo9D7GJ8G"
+      },
+      "source": [
+        "```\n",
+        "Copyright 2020 The IREE Authors\n",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:28]: Copyright year 2021

```suggestion
        "Copyright 2021 The IREE Authors\n",
```

I'm also tempted to switch these back to the collapsed license header format that we had before (copied from TF). WDYT?

before: 
image

after:
image

here's what TF does ([notebook](https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/quickstart/beginner.ipynb)):
image

which expands to this:
image

File: colab/tflite_text_classification.ipynb [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "T8WMt2sPS4ft"
+      },
+      "source": [
+        "%%capture\n",
+        "!python -m pip install iree-compiler-snapshot iree-runtime-snapshot iree-tools-tflite-snapshot -f https://github.com/google/iree/releases/snapshot-20210521.291\n",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:63]: let's set this to the latest release like the other notebooks
the "latest" tag suggestion on https://github.com/google/iree/issues/6046 may help keep it stable

File: colab/tflite_text_classification.ipynb [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "dweksnWs1eZb"
+      },
+      "source": [
+        "## Load the TFLite model"
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:99]: Since this is collapsed by default now, you could put more information in this text cell to answer "should I look at these details?"

image

Copying my earlier message:

> There should be enough context there that people can see if their own different models will need similar treatment, or if they can skip over that section entirely.

Suggested wording (feel free to go with something else):

```
## Load the TFLite model

* Download files for the pretrained model
* Extract model metadata used for input pre-processing and output post-processing
* Define helper functions for pre- and post-processing

These steps will differ from model to model. Consult the model source or reference documentation for details.
```

File: colab/tflite_text_classification.ipynb [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "fvxhY1W7X7Gf"
+      },
+      "source": [
+        "## TFLite\n",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:275]: could qualify this section header (and the "IREE" one below) so people can more easily see that they are independent of one another

```suggestion
        "## Run using TFLite\n",
```

File: colab/tflite_text_classification.ipynb [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "_FgwUBtm7Y5n"
+      },
+      "source": [
+        "## IREE\n",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:348]:
```suggestion
        "## Run using IREE\n",
```

File: colab/tflite_text_classification.ipynb [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Nv8x81S5eV5g"
+      },
+      "source": [
+        "interpreter = tflite.Interpreter(\n",
+        "      model_path=str(ARTIFACTS_DIR.joinpath(\"text_classification.tflite\")), \n",
+        "      num_threads=None)\n",
```
Comment (ScottTodd) [inline colab/tflite_text_classification.ipynb:292]: Can you omit the `num_threads=None`?
These samples don't include it: https://www.tensorflow.org/lite/guide/python#run_an_inference_using_tflite_runtime

File: colab/README.md [hunk 8]
```diff
@@ -37,6 +37,14 @@ model, pre-processes it for import, then compiles it using IREE
 
 [![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/google/iree/blob/main/colab/tensorflow_hub_import.ipynb)
 
+### [tflite_text_classificaiton\.ipynb](tflite_text_classification.ipynb)
```
Comment (ScottTodd) [inline colab/README.md:40]: (In the future, not necessarily on this PR)

Let's also link to this from https://google.github.io/iree/ml-frameworks/tensorflow-lite/ as part of https://github.com/google/iree/issues/5455

**Review by ScottTodd — APPROVED** [review]
Nice, thanks for the updates! A few minor comments then LGTM

File: colab/tflite_text_classification.ipynb (hunk 5 above)
Comment (KoolJBlack) [inline colab/tflite_text_classification.ipynb:275]: To add to scott's comments, I think you can make it clearer that the "TFLIte and "IREE" sections are in essence doing the same thing but with different runtimes. 

In addition to altering th etitles, you can give this section a bigger header to separate the "setup" section from the "runtime section" and follow with a note under the latter that the TFLite and IREE runtimes should generate the same output.

**Review by KoolJBlack — APPROVED** [review]
Approval after addressing comment about sections.

File: colab/tflite_text_classification.ipynb (hunk 2 above)
Comment (not-jenni) [reply colab/tflite_text_classification.ipynb:28]: I think it seems okay for now to leave the new header so it is consistent with the other colab notebooks, but in a followup PR I think it would be good to update all of them so that they are collapsed
